File: frontend/e2e/steps/feature-flag.steps.ts
import { Given, When, Then, expect } from '../fixtures/test-fixtures';

// Feature Flag Steps

Given(
  'a feature flag {string} exists with enabled {string}',
  async ({ apiClient, testContext, testSuffix }, name: string, enabled: string) => {
    await apiClient.loginAsAdmin();
    const uniqueName = `${name}${testSuffix}`;
    const flag = await apiClient.createFeatureFlag(
      uniqueName,
      enabled.toLowerCase() === 'true',
      `Test flag ${name}`
    );
    testContext.set(`featureFlag:${name}`, { ...flag, actualName: uniqueName });
    testContext.set(`featureFlagName:${name}`, uniqueName);
  }
);

Given(
  'a feature flag {string} exists and is disabled',
  async ({ apiClient, testContext, testSuffix }, name: string) => {
    await apiClient.loginAsAdmin();
    const uniqueName = `${name}${testSuffix}`;
    const flag = await apiClient.createFeatureFlag(
      uniqueName,
      false,
      `Test flag ${name}`
    );
    testContext.set(`featureFlag:${name}`, { ...flag, actualName: uniqueName });
    testContext.set(`featureFlagName:${name}`, uniqueName);
  }
);

Given(
  'a feature flag {string} exists and is enabled',
  async ({ apiClient, testContext, testSuffix }, name: string) => {
    await apiClient.loginAsAdmin();
    const uniqueName = `${name}${testSuffix}`;
    const flag = await apiClient.createFeatureFlag(
      uniqueName,
      true,
      `Test flag ${name}`
    );
    testContext.set(`featureFlag:${name}`, { ...flag, actualName: uniqueName });
    testContext.set(`featureFlagName:${name}`, uniqueName);
  }
);

When(
  'I create a feature flag {string} with description {string}',
  async ({ apiClient, testContext, testSuffix }, name: string, description: string) => {
    // Use API to create the feature flag (no admin UI exists for this)
    await apiClient.loginAsAdmin();
    const uniqueName = `${name}${testSuffix}`;
    const flag = await apiClient.createFeatureFlag(uniqueName, false, description);
    testContext.set(`featureFlag:${name}`, { ...flag, actualName: uniqueName });
    testContext.set(`featureFlagName:${name}`, uniqueName);
  }
);

When(
  'I enable the feature flag {string}',
  async ({ apiClient, testContext }, name: string) => {
    // Use API to enable the feature flag
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    await apiClient.updateFeatureFlag(actualName, { enabled: true });
  }
);

When(
  'I disable the feature flag {string}',
  async ({ apiClient, testContext }, name: string) => {
    // Use API to disable the feature flag
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    await apiClient.updateFeatureFlag(actualName, { enabled: false });
  }
);

When(
  'I delete the feature flag {string}',
  async ({ apiClient, testContext }, name: string) => {
    // Use API to delete the feature flag
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    await apiClient.deleteFeatureFlag(actualName);
  }
);

Then(
  'the feature flag {string} should exist',
  async ({ apiClient, testContext }, name: string) => {
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    const flags = await apiClient.getFeatureFlags();
    const exists = flags.results.some(f => f.name === actualName);
    expect(exists).toBe(true);
  }
);

Then(
  'the feature flag {string} should not exist',
  async ({ apiClient, testContext }, name: string) => {
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    const flags = await apiClient.getFeatureFlags();
    const exists = flags.results.some(f => f.name === actualName);
    expect(exists).toBe(false);
  }
);

Then(
  'the feature flag {string} should have description {string}',
  async ({ apiClient, testContext }, name: string, expectedDesc: string) => {
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    const flag = await apiClient.getFeatureFlag(actualName);
    expect(flag).toBeDefined();
  }
);

Then(
  'the feature flag {string} should be enabled',
  async ({ apiClient, testContext }, name: string) => {
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    const flag = await apiClient.getFeatureFlag(actualName);
    expect(flag.enabled).toBe(true);
  }
);

Then(
  'the feature flag {string} should be disabled',
  async ({ apiClient, testContext }, name: string) => {
    await apiClient.loginAsAdmin();
    const actualName = testContext.get(`featureFlagName:${name}`) as string || name;
    const flag = await apiClient.getFeatureFlag(actualName);
    expect(flag.enabled).toBe(false);
  }
);

File: backend/tests/bdd/steps/feature_flag_steps.py
"""Feature flag-related step definitions."""

from typing import Any, Dict

from pytest_bdd import given, when, then, parsers

from application.services.feature_flag_service import FeatureFlagService
from domain.user_context import UserContext


# ============================================================================
# GIVEN steps
# ============================================================================


@given(parsers.parse('a feature flag "{name}" exists and is disabled'))
def feature_flag_exists_disabled(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Create a disabled feature flag."""
    flag = feature_flag_service.create(
        name=name,
        enabled=False,
        description="",
        user_context=admin_user_context,
    )
    context["feature_flags"][name] = flag


@given(parsers.parse('a feature flag "{name}" exists and is enabled'))
def feature_flag_exists_enabled(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Create an enabled feature flag."""
    flag = feature_flag_service.create(
        name=name,
        enabled=True,
        description="",
        user_context=admin_user_context,
    )
    context["feature_flags"][name] = flag


# ============================================================================
# WHEN steps
# ============================================================================


@when(parsers.parse('I create a feature flag "{name}" with description "{description}"'))
def create_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
    description: str,
) -> None:
    """Create a feature flag (expected to succeed)."""
    flag = feature_flag_service.create(
        name=name,
        enabled=False,
        description=description,
        user_context=context["current_user_context"],
    )
    context["feature_flags"][name] = flag


@when(parsers.parse('I try to create a feature flag "{name}" with description "{description}"'))
def try_create_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
    description: str,
) -> None:
    """Try to create a feature flag (may fail)."""
    try:
        flag = feature_flag_service.create(
            name=name,
            enabled=False,
            description=description,
            user_context=context["current_user_context"],
        )
        context["feature_flags"][name] = flag
        context["error"] = None
    except Exception as e:
        context["error"] = e


@when(parsers.parse('I enable the feature flag "{name}"'))
def enable_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
) -> None:
    """Enable a feature flag (expected to succeed)."""
    flag = feature_flag_service.update(
        flag_name=name,
        enabled=True,
        description=None,
        user_context=context["current_user_context"],
    )
    context["feature_flags"][name] = flag


@when(parsers.parse('I try to enable the feature flag "{name}"'))
def try_enable_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
) -> None:
    """Try to enable a feature flag (may fail)."""
    try:
        flag = feature_flag_service.update(
            flag_name=name,
            enabled=True,
            description=None,
            user_context=context["current_user_context"],
        )
        context["feature_flags"][name] = flag
        context["error"] = None
    except Exception as e:
        context["error"] = e


@when(parsers.parse('I disable the feature flag "{name}"'))
def disable_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
) -> None:
    """Disable a feature flag (expected to succeed)."""
    flag = feature_flag_service.update(
        flag_name=name,
        enabled=False,
        description=None,
        user_context=context["current_user_context"],
    )
    context["feature_flags"][name] = flag


@when(parsers.parse('I delete the feature flag "{name}"'))
def delete_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
) -> None:
    """Delete a feature flag (expected to succeed)."""
    feature_flag_service.delete(
        flag_name=name,
        user_context=context["current_user_context"],
    )
    if name in context["feature_flags"]:
        del context["feature_flags"][name]


@when(parsers.parse('I try to delete the feature flag "{name}"'))
def try_delete_feature_flag(
    context: Dict[str, Any],
    feature_flag_service: FeatureFlagService,
    name: str,
) -> None:
    """Try to delete a feature flag (may fail)."""
    try:
        feature_flag_service.delete(
            flag_name=name,
            user_context=context["current_user_context"],
        )
        if name in context["feature_flags"]:
            del context["feature_flags"][name]
        context["error"] = None
    except Exception as e:
        context["error"] = e


# ============================================================================
# THEN steps
# ============================================================================


@then(parsers.parse('the feature flag "{name}" should exist'))
def feature_flag_should_exist(
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Verify a feature flag exists."""
    flags = feature_flag_service.get_all(admin_user_context)
    assert any(f.name == name for f in flags), f"Feature flag '{name}' not found"


@then(parsers.parse('the feature flag "{name}" should not exist'))
def feature_flag_should_not_exist(
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Verify a feature flag does not exist."""
    flags = feature_flag_service.get_all(admin_user_context)
    assert not any(f.name == name for f in flags), f"Feature flag '{name}' should not exist"


@then(parsers.parse('the feature flag "{name}" should be enabled'))
def feature_flag_should_be_enabled(
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Verify a feature flag is enabled."""
    flag = feature_flag_service.get_by_name(name, admin_user_context)
    assert flag is not None, f"Feature flag '{name}' not found"
    assert flag.enabled, f"Feature flag '{name}' should be enabled"


@then(parsers.parse('the feature flag "{name}" should be disabled'))
def feature_flag_should_be_disabled(
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
) -> None:
    """Verify a feature flag is disabled."""
    flag = feature_flag_service.get_by_name(name, admin_user_context)
    assert flag is not None, f"Feature flag '{name}' not found"
    assert not flag.enabled, f"Feature flag '{name}' should be disabled"


@then(parsers.parse('the feature flag "{name}" should have description "{expected_desc}"'))
def feature_flag_should_have_description(
    feature_flag_service: FeatureFlagService,
    admin_user_context: UserContext,
    name: str,
    expected_desc: str,
) -> None:
    """Verify a feature flag has the expected description."""
    flag = feature_flag_service.get_by_name(name, admin_user_context)
    assert flag is not None, f"Feature flag '{name}' not found"
    assert flag.description == expected_desc, (
        f"Expected description '{expected_desc}', got '{flag.description}'"
    )
